src/backend/optimizer/util/plancat.c

```diff
@@ -80,6 +83,39 @@ static void set_baserel_partition_key_exprs(Relation relation,
 static void set_baserel_partition_constraint(Relation relation,
 											 RelOptInfo *rel);
 
+static bool
+is_index_valid(Relation index, LOCKMODE lmode)
+{
+	if (!index->rd_index->indisvalid)
+		return false;
+
+	if (index->rd_rel->relpersistence == RELPERSISTENCE_UNLOGGED)
+	{
+		while (true)
+		{
+			Buffer metapage = ReadBuffer(index, 0);
+			bool isNew = PageIsNew(BufferGetPage(metapage));
+			ReleaseBuffer(metapage);
+			if (isNew)
+			{
+				Relation heap;
+				if (lmode != ExclusiveLock)
+				{
+					UnlockRelation(index, lmode);
+					LockRelation(index, ExclusiveLock);
+					lmode = ExclusiveLock;
+					continue;
```

Comment:
Why? We need to block table if we are adding or removing index.
But please notice that we do not actually create index in this case - it is already exists.
We just reinitialize it. 
And we need to prevent any other manipulations with index at this moment, but not with heap.

Reply:
Sorry, I do not understand.
Any newly inserted records will be visible for other transactions only when this transaction is completed and new record is inserted in all indexes. But index is locked and so insert should wait until it is reinitialized.

Actually there is no any difference of index reinitialization with any other update operation with index (insert, delete, ...). We do not  need to lock table to perform such operation.

Reply:
IOW, this case is different from initial index creation because in the latter index tuple and index itself is not visible to other backends till the end of transaction, so to prevent insertions table itself is locked; but here index already exists, right?

Reply:
Yes, exactly. When index is created or dropped lock of the table is needed to avoid situation when record is inserted in heap but not in index or visa versa. But in our case index is already present and so Postgres will always include inserted record in the index. And if index is locked, then insert operation will wait until lock is released.